### PR TITLE
Job: Add evaluation step comments in the syncJob

### PR DESCRIPTION
**Pull request by tenzen-y**
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:
I added the evaluation steps comment in the syncJob function to clarify the evaluation orders.

#### Which issue(s) this PR fixes:

Part-of https://github.com/kubernetes/kubernetes/issues/128564

#### Special notes for your reviewer:
I would like to focus on controller comments in this PR since this is not needed for the API review.

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: pkg/controller/job/job_controller.go [hunk 1]
```diff
@@ -944,10 +957,14 @@ func (jm *Controller) syncJob(ctx context.Context, key string) (rErr error) {
 			}
 			jobCtx.podsWithDelayedDeletionPerIndex = getPodsWithDelayedDeletionPerIndex(logger, jobCtx)
 		}
-		if jobCtx.finishedCondition == nil {
-			if msg, met := matchSuccessPolicy(logger, job.Spec.SuccessPolicy, *job.Spec.Completions, jobCtx.succeededIndexes); met {
-				jobCtx.finishedCondition = newCondition(batch.JobSuccessCriteriaMet, v1.ConditionTrue, batch.JobReasonSuccessPolicy, msg, jm.clock.Now())
-			}
```
Comment (tenzen-y) [inline pkg/controller/job/job_controller.go:950]: I'm on the fence whether or not I should revert this change. But when reverting this, 2.3 and 3 comments might look like mixed.

Any thoughts?

Comment (mimowo) [reply pkg/controller/job/job_controller.go:950]: TBH, I would avoid changing the production code just for the  comments.

File: pkg/controller/job/job_controller.go [hunk 2]
```diff
@@ -902,9 +902,20 @@ func (jm *Controller) syncJob(ctx context.Context, key string) (rErr error) {
 
 	var manageJobErr error
 
+	// This is the starting point for the Jobs evaluation steps.
+	// As a Job any kind of criteria evaluation principal, the negative ones (e.g., .spec.backoffLimit, and .spec.activeDeadlineSeconds)
+	// are respected rather than the positive ones (e.g., .spec.completions, and .spec.successPolicy).
+	// And the Job completion (succeeded) can be observed by the existence of the completionTime (.status.completionTime)
+	// instead of succeeded state (.status.succeeded).
+	//
+	// 1. Preserve the existing Job state.
```
Comment (mimowo) [inline pkg/controller/job/job_controller.go:911]: This seems quite verbose and reduces the chance someone will actually read it. 
Maybe we can have something more light weight, like:

```
This is the starting point for evaluating the end state of the Job. 
Note that in some cases a Job could satisfy multiple checks at the same time, and so we need to order them. 
We implement the following order of checks:
1. pre-existing SuccessCriteriaMet and FailureTarget are read and respected
2. failure scenarios (ActiveDeadlineSeconds exceeded, BackoffLimitExceeded, PodFailurePolicy, etc.)
3. success scenarios (Completions reached or SuccessPolicy met)
```
I also don't have a strong opinion here, but I don't see value for enumerating the checks like "1", "2.1", etc. It adds overhead on maintenance in the future, and it is pretty obvious what the code does. In my view it is more important to capture the overall decision (as above) rather than annotate code line-by-line.

wdyt @atiratree @soltysh ?

File: pkg/controller/job/job_controller.go (hunk 1 above)
Comment (soltysh) [reply pkg/controller/job/job_controller.go:950]: I'm with Michal here, let's leave it as is.

File: pkg/controller/job/job_controller.go (hunk 2 above)
Comment (soltysh) [reply pkg/controller/job/job_controller.go:911]: I agree the intro here could be shorter, something like quick:
```
1. pre-existing SuccessCriteriaMet and FailureTarget are read and respected
2. failure scenarios
3. success scenarios
```
than in the appropriate places you can describe exactly which if stands for which failure case. See my comments below.

File: pkg/controller/job/job_controller.go [hunk 3]
```diff
@@ -916,6 +927,7 @@ func (jm *Controller) syncJob(ctx context.Context, key string) (rErr error) {
 			jobCtx.finishedCondition = newCondition(batch.JobFailureTarget, v1.ConditionTrue, batch.JobReasonPodFailurePolicy, *failJobMessage, jm.clock.Now())
 		}
 	}
+	// 2.2. Evaluate any other negative criteria.
```
Comment (soltysh) [inline pkg/controller/job/job_controller.go:930]: Be specific here about what those criteria are, here backoffLimitExceeded, activeDeadlineExceeded.

File: pkg/controller/job/job_controller.go [hunk 4]
```diff
@@ -930,6 +942,7 @@ func (jm *Controller) syncJob(ctx context.Context, key string) (rErr error) {
 		}
 	}
 
+	// 2.3. Evaluate negative criteria only for the Indexed Jobs, and calculate the succeeded Indexes.
```
Comment (soltysh) [inline pkg/controller/job/job_controller.go:945]: Again, add more info about what specific negative criteria are mentioned here.

File: pkg/controller/job/job_controller.go (hunk 2 above)
Comment (mimowo) [reply pkg/controller/job/job_controller.go:911]: This sgtm too

Comment (tenzen-y) [reply pkg/controller/job/job_controller.go:911]: That makes sense. Let me make this shorter, then add more details to each block.

File: pkg/controller/job/job_controller.go (hunk 1 above)
Comment (tenzen-y) [reply pkg/controller/job/job_controller.go:950]: That makes sense. Let me revert this

File: pkg/controller/job/job_controller.go (hunk 2 above)
Comment (atiratree) [reply pkg/controller/job/job_controller.go:911]: +1 to the above and making sure the comments don't drift over time. Having the prefix in the same form like we do now `Evaluate the negative criteria.` should be enough instead of the numbering. Or just `Negative Criteria: X` ?

File: pkg/controller/job/job_controller.go (hunk 3 above)
Comment (tenzen-y) [reply pkg/controller/job/job_controller.go:930]: That makes sense, thanks!

File: pkg/controller/job/job_controller.go (hunk 2 above)
Comment (tenzen-y) [reply pkg/controller/job/job_controller.go:911]: > Having the prefix in the same form like we do now Evaluate the negative criteria. should be enough instead of the numbering. 

I'm leaning toward numbering since I believe that breaking the ordering implies breaking specifications or that the external developer needs to change the ordering in their controllers.

So, I think the numbering would be better for finding the changed evaluation order in the future. 
@atiratree WDYT?

File: pkg/controller/job/job_controller.go (hunk 4 above)
Comment (tenzen-y) [reply pkg/controller/job/job_controller.go]: Thanks, I addressed.

File: pkg/controller/job/job_controller.go (hunk 3 above)
Comment (tenzen-y) [reply pkg/controller/job/job_controller.go]: Done.

File: pkg/controller/job/job_controller.go (hunk 2 above)
Comment (tenzen-y) [reply pkg/controller/job/job_controller.go]: Done.

File: pkg/controller/job/job_controller.go (hunk 1 above)
Comment (tenzen-y) [reply pkg/controller/job/job_controller.go]: Done.

**Comment by tenzen-y** [comment]
@soltysh @atiratree @mimowo Thanks for the feedback. I addressed all comments, please take another look.

File: pkg/controller/job/job_controller.go [hunk 5]
```diff
@@ -902,9 +902,19 @@ func (jm *Controller) syncJob(ctx context.Context, key string) (rErr error) {
...
+	// 1. Evaluate the pre-existing SuccessCriteriaMet and FailureTarget to respect the previous reconcile results.
+	// 2. Evaluate failure scenarios
+	// 3. Evaluate success scenarios
+
 	exceedsBackoffLimit := jobCtx.failed > *job.Spec.BackoffLimit
+	// Evaluate the pre-existing SuccessCriteriaMet.
 	jobCtx.finishedCondition = hasSuccessCriteriaMetCondition(&job)
 
+	// Evaluate the pre-existing FailureTarget.
+	// After that evaluate failure scenarios for the PodFailurePolicy only when the Job does not have FailureTarget.
```
Comment (mimowo) [inline pkg/controller/job/job_controller.go:916]: nit: I think this sentence is redundant, it also feels inconsistent - we don't add the sentence wrt to SuccessCriteriaMet. So I propose to drop it completely.

**Review by mimowo — COMMENTED** [review]
LGTM, just one nit

File: pkg/controller/job/job_controller.go (hunk 2 above)
Comment (mimowo) [reply pkg/controller/job/job_controller.go]: sgtm

File: pkg/controller/job/job_controller.go (hunk 5 above)
Comment (tenzen-y) [reply pkg/controller/job/job_controller.go:916]: Actually, there are a slightly different between SuccessCriteriaMet and FailureTarget evaluation like below:

- SuccessCriteriaMet
  - we evaluate the pre-existing one and new one in separate places (pre-existing is Line 912, new one is Line 959)
- FailureTarget
  - we evaluate both in here (Line 920 - 927)

Comment (tenzen-y) [reply pkg/controller/job/job_controller.go:916]: So, pre-existing and new one evaluations for FailureTarget are combined to here.

Comment (mimowo) [reply pkg/controller/job/job_controller.go:916]: I see, in that case, I would suggest to 
1. rephrase. for consistency, the comment "// After that evaluate failure scenarios for the PodFailurePolicy only when the Job does not have FailureTarget." as "Evaluate the failure scenarios for the PodFailurePolicy." 
1. move both comments to L921, just before "failureTargetCondition := findConditionByType(job.Status.Conditions, batch.JobFailureTarget)".

Comment (tenzen-y) [reply pkg/controller/job/job_controller.go]: That makes sense.
Done.

File: pkg/controller/job/job_controller.go [hunk 6]
```diff
@@ -902,20 +902,30 @@ func (jm *Controller) syncJob(ctx context.Context, key string) (rErr error) {
...
 	jobCtx.finishedCondition = hasSuccessCriteriaMetCondition(&job)
 
 	// Given that the Job already has the SuccessCriteriaMet condition, the termination condition already had confirmed in another cycle.
 	// So, the job-controller evaluates the podFailurePolicy only when the Job doesn't have the SuccessCriteriaMet condition.
 	if jobCtx.finishedCondition == nil {
+		// Evaluate the pre-existing FailureTarget.
 		failureTargetCondition := findConditionByType(job.Status.Conditions, batch.JobFailureTarget)
 		if failureTargetCondition != nil && failureTargetCondition.Status == v1.ConditionTrue {
 			jobCtx.finishedCondition = newFailedConditionForFailureTarget(failureTargetCondition, jm.clock.Now())
+			// Evaluate failure scenarios for the PodFailurePolicy.
```
Comment (tenzen-y) [inline pkg/controller/job/job_controller.go:922]: This head of the weird tab brought by the `go fmt`.

Comment (mimowo) [inline pkg/controller/job/job_controller.go:922]:
```suggestion
			// Evaluate failure scenarios for PodFailurePolicy.
```
Disclaimer: not an English expert here, but I think PodFailurePolicy is a proper noun in this context, so we can skip "the". Same for other feature names below.

File: pkg/controller/job/job_controller.go [hunk 7]
```diff
@@ -933,6 +943,7 @@ func (jm *Controller) syncJob(ctx context.Context, key string) (rErr error) {
 	if isIndexedJob(&job) {
 		jobCtx.prevSucceededIndexes, jobCtx.succeededIndexes = calculateSucceededIndexes(logger, &job, pods)
 		jobCtx.succeeded = int32(jobCtx.succeededIndexes.total())
+		// Evaluate the failure scenarios for the BackoffLimitPerIndex.
```
Comment (mimowo) [inline pkg/controller/job/job_controller.go:946]:
```suggestion
		// Evaluate failure scenarios for BackoffLimitPerIndex.
```

File: pkg/controller/job/job_controller.go [hunk 8]
```diff
@@ -971,6 +983,7 @@ func (jm *Controller) syncJob(ctx context.Context, key string) (rErr error) {
 			active, action, manageJobErr = jm.manageJob(ctx, &job, jobCtx)
 			manageJobCalled = true
 		}
+		// Evaluate the success scenarios for the Completions.
```
Comment (mimowo) [inline pkg/controller/job/job_controller.go:986]:
```suggestion
		// Evaluate success scenarios for Completions.
```

File: pkg/controller/job/job_controller.go (hunk 6 above)
Comment (tenzen-y) [reply pkg/controller/job/job_controller.go:922]: I'm also not well familiar with English :) So, I have no preference or opinion on this.

Comment (mimowo) [reply pkg/controller/job/job_controller.go:922]: in that case, maybe apply the suggestion, but feel free to change back if others request.

File: pkg/controller/job/job_controller.go (hunk 8 above)
Comment (tenzen-y) [reply pkg/controller/job/job_controller.go:986]: What is the motivation for using the Reason only here?

Comment (mimowo) [reply pkg/controller/job/job_controller.go:986]: I thought completions alone may not be descriptive, but it makes more sense actually as in other comments you use the fields from spec

Comment (mimowo) [reply pkg/controller/job/job_controller.go:986]: So keep as it was

File: pkg/controller/job/job_controller.go (hunk 6 above)
Comment (tenzen-y) [reply pkg/controller/job/job_controller.go]: Sure, makes sense.
Done.

File: pkg/controller/job/job_controller.go [hunk 9]
```diff
@@ -902,20 +902,30 @@ func (jm *Controller) syncJob(ctx context.Context, key string) (rErr error) {
...
 		failureTargetCondition := findConditionByType(job.Status.Conditions, batch.JobFailureTarget)
 		if failureTargetCondition != nil && failureTargetCondition.Status == v1.ConditionTrue {
 			jobCtx.finishedCondition = newFailedConditionForFailureTarget(failureTargetCondition, jm.clock.Now())
+			// Evaluate failure scenarios for the PodFailurePolicy.
 		} else if failJobMessage := getFailJobMessage(&job, pods); failJobMessage != nil {
 			// Prepare the interim FailureTarget condition to record the failure message before the finalizers (allowing removal of the pods) are removed.
 			jobCtx.finishedCondition = newCondition(batch.JobFailureTarget, v1.ConditionTrue, batch.JobReasonPodFailurePolicy, *failJobMessage, jm.clock.Now())
 		}
 	}
+	// Evaluate the failure scenarios for the BackoffLimit and the ActiveDeadlineSeconds.
```
[the remark this answers is not in the document]
Comment (tenzen-y) [reply pkg/controller/job/job_controller.go]: Done.

File: pkg/controller/job/job_controller.go (hunk 7 above)
Comment (tenzen-y) [reply pkg/controller/job/job_controller.go]: Done.

File: pkg/controller/job/job_controller.go [hunk 10]
```diff
@@ -944,6 +955,7 @@ func (jm *Controller) syncJob(ctx context.Context, key string) (rErr error) {
 			}
 			jobCtx.podsWithDelayedDeletionPerIndex = getPodsWithDelayedDeletionPerIndex(logger, jobCtx)
 		}
+		// Evaluate the success scenarios for the SuccessPolicy.
```
[the remark this answers is not in the document]
Comment (tenzen-y) [reply pkg/controller/job/job_controller.go]: Done.

File: pkg/controller/job/job_controller.go (hunk 8 above)
Comment (tenzen-y) [reply pkg/controller/job/job_controller.go]: Thank you for checking that.

**Review by mimowo — COMMENTED** [review]
/lgtm
/assign @atiratree @soltysh

File: pkg/controller/job/job_controller.go [hunk 11]
```diff
@@ -902,20 +902,30 @@ func (jm *Controller) syncJob(ctx context.Context, key string) (rErr error) {
 
 	var manageJobErr error
 
+	// This is the starting point for evaluating the end state of the Job.
+	// Note that we need to order evaluations since a Job could satisfy multiple criteria at the same time in some cases:
+	// 1. Evaluate the pre-existing SuccessCriteriaMet and FailureTarget to respect the previous reconcile results.
+	// 2. Evaluate failure scenarios
+	// 3. Evaluate success scenarios
```
[the remark this answers is not in the document]
Comment (mimowo) [reply pkg/controller/job/job_controller.go:909]: Yeah, I don't think there is any technical reason for the different places of promoting the conditions. So I think aligning that would be a great follow up cleanup. Since success policy is still beta while failure policy is ga, maybe we could align into that order of things, but I'm ok either way   WDYT?

[the remark this answers is not in the document]
Comment (tenzen-y) [reply pkg/controller/job/job_controller.go:909]: @atiratree Does that mean like the comments below?

```
	// This is the starting point for evaluating the end state of the Job.
	// Note that we need to order evaluations since a Job could satisfy multiple criteria at the same time in some cases:
	// 1. Evaluate the pre-existing SuccessCriteriaMet and FailureTarget to respect the previous reconcile results.
	// 2. Evaluate failure scenarios
	// 3. Evaluate success scenarios
	// 4. Evaluate finishedCondition, then transform FailureTarget to Failed and SuccessCriteriaMet to Complete.
```

[the remark this answers is not in the document]
Comment (atiratree) [reply pkg/controller/job/job_controller.go:909]: Ok, we can revisit this with the success policy (or even separately). It would be nice to unify these to decrease the cognitive load. Let's at least document it for now.

[the remark this answers is not in the document]
Comment (tenzen-y) [reply pkg/controller/job/job_controller.go:909]: > FailureTarget (transformed to Failed here and later)
SuccessCriteriaMet (transformed to Complete later)

Ah, I found the differences. FailureTarget and SuccessPolicy are transformed in different place

[the remark this answers is not in the document]
Comment (tenzen-y) [reply pkg/controller/job/job_controller.go:909]:
```
[...]
	// 1. Evaluate the pre-existing SuccessCriteriaMet and FailureTarget to respect the previous reconcile results. FailureTarget is transformed to Failed condition.
[...]
```

@atiratree How about this?

[the remark this answers is not in the document]
Comment (tenzen-y) [reply pkg/controller/job/job_controller.go:909]: That makes sense.
It looks like both are valuable. So, let me refine `1` and add `4`.

[the remark this answers is not in the document]
Comment (tenzen-y) [reply pkg/controller/job/job_controller.go]: Done. This is the differeneces: https://github.com/kubernetes/kubernetes/compare/72791344060005fc4b99016893c28e44953e0f08..5dda60ee4ef4769dc63dcd312a5cbb0c3023df29

**Comment by atiratree** [comment]
/lgtm

@mimowo @soltysh for the new updates

**Review by soltysh — APPROVED** [review]
/approve

**Comment by mimowo** [comment]
LGTM

File: pkg/controller/job/job_controller.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (tenzen-y) [reply pkg/controller/job/job_controller.go]: @atiratree For refactoring the SuccessPolicy transformation, I would like to work on it after the 1.32 code freeze since I think the status transition refactoring should be done carefully.

[the remark this answers is not in the document]
Comment (atiratree) [reply pkg/controller/job/job_controller.go]: sure, there is no rush with that
